availability/etcdfailover/electionmanager.go

```
@@ -0,0 +1,234 @@
+		em.StopAndLog()
+		return nil, nil, err
+	}
+
+	errChan := make(chan error)
+	go func(runnable func() error) {
+		em.cfg.Logger.Debug("waiting for election event or quit signal")
+		for evt := range electionChan {
+			switch evt {
+			case Participant:
+				em.cfg.Logger.Info("tracker is eligible to become leader")
+			case Leader:
+				em.cfg.Logger.Info("leader among trackers")
+				if err = runnable(); err != nil {
+					em.cfg.Logger.Errorf("encountered error in runnable: %v", err)
+					errChan <- err
+				}
+			}
+		}
+		errChan <- fmt.Errorf("failed in election")
+	}(runnable)
+
+	return errChan, em.Stop, nil
+}
+
+//method not thread-safe, ensure it is called after acquiring mutex of election manager
+func (em *ElectionManager) stop() error {
+	if em.state != NotParticipant {
+		em.cfg.Logger.Infof("stopping election manager in state: %s", em.state)
+		//to ensure consecutive cleanups don't happen
+		defer func() {
+			em.state = NotParticipant
+		}()
+		em.evtChan.Close()
+
+		// check for nil on election, session and cli because its possible that `ElectionManager.Start()` failed somewhere in between
+		if em.election != nil {
+			em.cfg.Logger.Debug("resigning from election")
+			// wait on resign for at least session timeout to ensure that if resign fails in case of n/w partition or etcd disaster, session has reliably expired
+			cctx, _ := context.WithTimeout(context.Background(), time.Duration(em.cfg.SessionTimeoutSecs) * time.Second)
+			err := em.election.Resign(cctx)
+			if err != nil {
+				em.cfg.Logger.Warnf("error resigning from election: %v", err)
+			}
+		}
+		em.cancel()
+
+		if em.sessionCreated && em.session != nil {
+			em.cfg.Logger.Debug("closing etcd session")
+			if err := em.session.Close(); err != nil {
+				em.cfg.Logger.Warnf("error closing session: v%", err.Error())
+			}
+		}
+
+		if em.cliCreated && em.cli != nil {
+			em.cfg.Logger.Debug("closing etcd client")
+			if err := em.cli.Close(); err != nil {
+				return err
```

any reason not to return session close or Resign error?

---

it will anyways resign since we wait till session timeout. I don't want to return an error and calling believing that since stop failed, I am still a participant.
If you see session.close, it does something similar to ensure lease is expired.